src/coreclr/jit/objectalloc.cpp

```diff
@@ -1335,3 +1483,1734 @@ void ObjectAllocator::RewriteUses()
+        {
+            JITDUMP("... not BBJ_COND\n");
+            continue;
+        }
+
+        // We require that one idomBlock successor *not* dominate.
+        // (otherwise idomBlock could be the top of a diamond where both outcomes reach block).
+        //
+        const bool trueSuccessorDominates  = comp->m_domTree->Dominates(idomBlock->GetTrueTarget(), block);
+        const bool falseSuccessorDominates = comp->m_domTree->Dominates(idomBlock->GetFalseTarget(), block);
+
+        if (trueSuccessorDominates && falseSuccessorDominates)
+        {
+            JITDUMP("... both successors dominate?\n");
+            continue;
+        }
+
+        if (!(trueSuccessorDominates || falseSuccessorDominates))
+        {
+            JITDUMP("... neither successor dominates\n");
+            continue;
+        }
+
+        GenTree* const guardingRelop = IsGuard(idomBlock, info);
+        if (guardingRelop == nullptr)
+        {
+            continue;
+        }
+
+        // We found a dominating GDV test, see if the condition is the one we're looking for.
+        //
+        if (testOutcome)
+        {
+            bool const isReachableOnGDVSuccess = (trueSuccessorDominates && guardingRelop->OperIs(GT_EQ)) ||
+                                                 (falseSuccessorDominates && guardingRelop->OperIs(GT_NE));
+            if (isReachableOnGDVSuccess)
+            {
+                info->m_block = idomBlock;
+                return true;
+            }
+            JITDUMP("... guarded by failing GDV\n");
+            continue;
+        }
+        else
+        {
+            bool const isReachableOnGDVFailure = (trueSuccessorDominates && guardingRelop->OperIs(GT_NE)) ||
+                                                 (falseSuccessorDominates && guardingRelop->OperIs(GT_EQ));
+            if (isReachableOnGDVFailure)
+            {
+                info->m_block = idomBlock;
+                return true;
+            }
+            JITDUMP("... guarded by successful GDV\n");
+            continue;
+        }
+    }
+
+    JITDUMP("... no more doms\n");
+    return false;
+}
+
+//------------------------------------------------------------------------------
+// IsGuard: does block look like a GDV guard
+//
+// Arguments:
+//   block -- block in question
+//   info -- [out] guard info
+//
+// Returns:
+//   Comparison tree if this is a guard, or nullptr
+//
+GenTree* ObjectAllocator::IsGuard(BasicBlock* block, GuardInfo* info)
+{
+    if (!block->KindIs(BBJ_COND))
+    {
+        JITDUMP("... not BBJ_COND\n");
+        return nullptr;
+    }
+
+    Statement* const stmt = block->lastStmt();
+    if (stmt == nullptr)
+    {
+        JITDUMP("... no stmt\n");
+        return nullptr;
+    }
+
+    GenTree* const jumpTree = stmt->GetRootNode();
+    if (!jumpTree->OperIs(GT_JTRUE))
+    {
+        JITDUMP("... no JTRUE\n");
+        return nullptr;
+    }
+
+    GenTree* const tree = jumpTree->AsOp()->gtOp1;
```

Comment:
Nit (here and also below)

```suggestion
    GenTree* const tree = jumpTree->gtGetOp1();
```